geoportailv3/static/js/layerpermalinkservice.js

```diff
@@ -0,0 +1,254 @@
+  goog.array.forEach(layers.getArray(), function(layer) {
+    goog.events.listen(layer, ol.ObjectEventType.PROPERTYCHANGE,
+        goog.bind(function() {
+          this.setLayerState_(selectedLayers);
+        }, this));
+  }, this);
+  // Add event listener to all future selected layers
+  goog.events.listen(layers, ol.CollectionEventType.ADD,
+      goog.bind(
+      /**
+       * @param {ol.CollectionEventType} collEvt Collection event.
+       */
+      function(collEvt) {
+        var layer = /** @type {ol.layer.Layer} */ (collEvt.element);
+        goog.events.listen(layer, ol.ObjectEventType.PROPERTYCHANGE,
+            goog.bind(
+            /**
+             * @param {ol.ObjectEventType} layerEvt Layer event
+             */
+            function(layerEvt) {
+              this.setLayerState_(selectedLayers);
+            }, this));
+      }, this)
+  );
+};
+
+
+/**
+ * @param {Array} element
+ * @return {Array} array
+ * @private
+ */
+app.LayerPermalinkManager.getAllChildren_ = function(element) {
+  var array = [];
+  for (var i = 0; i < element.length; i++) {
+    if (element[i].hasOwnProperty('children')) {
+      goog.array.extend(array, app.LayerPermalinkManager.getAllChildren_(
+          element[i].children)
+      );
+    } else {
+      element[i].id = element[i].id;
+      array.push(element[i]);
+    }
+  }
+  return array;
+};
+
+
+/**
+ * @param {angular.Scope} scope
+ * @param {ol.Map} map
+ * @param {Array.<ol.layer.Layer>} selectedLayers
+ */
+app.LayerPermalinkManager.prototype.init =
+    function(scope, map, selectedLayers) {
+  /**
+   * @type {angular.Scope}
+   */
+  this.scope_ = scope;
+
+  /**
+   * @type {ol.Map}
+   * @private
+   */
+  this.map_ = map;
+
+  var layerIds = /** @type {Array.<number>} */
+      (this.getStateValue_('layers'));
+
+  // Wait for themes to load before adding layers from state
+  goog.events.listen(this.appThemes_, app.ThemesEventType.LOAD,
+      /**
+       * @param {goog.events.Event} evt Event.
+       */
+      function(evt) {
+        this.appThemes_.getBgLayers().then(goog.bind(
+            /**
+             * @param {Array.<ol.layer.Base>} bgLayers
+             */
+            function(bgLayers) {
+              var stateLayerLabel =
+                  this.stateManager_.getInitialValue('bgLayer');
+              if (goog.isDefAndNotNull(stateLayerLabel)) {
+                var layer = /** @type {ol.layer.Base} */
+                    (goog.array.find(bgLayers, function(layer) {
+                  return layer.get('label') === stateLayerLabel;
+                }));
+                this.backgroundLayerMgr_.set(this.map_, layer);
+              }
+            }, this));
+
+        this.appThemes_.getThemesObject().then(
+            goog.bind(function(themes) {
+              if (goog.isDef(layerIds)) {
+                this.applyLayerStateToMap_(layerIds, themes);
+              }
+              this.setupWatchers_(selectedLayers);
```

Comment:
If you use `listenOnce` it won't work in the case where the user loads the application and then logs in, as you won't see the second `themes` response.